XenAdmin/Commands/RebootHostCommand.cs

```diff
@@ -120,12 +121,32 @@ protected override string ConfirmationDialogText
             get
             {
                 List<Host> hosts = GetSelection().AsXenObjects<Host>();
-                bool hasRunningVMs = (hosts.Find(h => h.resident_VMs.Count >= 2) != null);  // 2 not 1, because the Control Domain doesn't count
 
-                if (hosts.Count == 1)
-                    return (hasRunningVMs ? Messages.CONFIRM_REBOOT_SERVER : Messages.CONFIRM_REBOOT_SERVER_NO_VMS);
-                else
-                    return (hasRunningVMs ? Messages.CONFIRM_REBOOT_SERVERS : Messages.CONFIRM_REBOOT_SERVERS_NO_VMS);
+                bool hasRunningVMs = false;
+                var hciHosts = new List<Host>();
+
+                foreach (Host h in hosts)
+                {
+                    if (h.HasRunningVMs)
+                        hasRunningVMs = true;
+
+                    if (h.Connection.ResolveAll(h.resident_VMs).FindAll(v => v.HciWarnBeforeShutdown).Count > 0)

```

Comment:
I think Exists() would be better (instead of FindAll and Count>0), because it returns when a match is found ant that's all we need here.